View sent messages (further new database, add maillists)

Athozus opened the pull request:
Add the ability to view the previous sent messages. As it navigates through the messages of the player's contacts to then check if the sender is the player, I presume it cannot shows messages where all receivers are deleted from the contacts list. It adds a tabs widget, then I reduced the table height of 1 unit to let the place for the tabs.

fixes #14, #7 and #3

Looks like that 
![screenshot_20230223_113648](https://user-images.githubusercontent.com/75171564/220885900-7c9f6a10-0843-4903-b8b1-85d331ea3f88.png)

fluxionary reviewed (commented):
this doesn't work the way i'd expect it to.

for one, i cannot read messages i've sent. secondly, if the recipient (when there is a single recipient) deletes the message, then i no longer see it in my "sent" list. i'd personally much prefer something that worked like the "sent" box in actual email systems.

S-S-X commented:
> for one, i cannot read messages i've sent. secondly, if the recipient (when there is a single recipient) deletes the message, > then i no longer see it in my "sent" list. i'd personally much prefer something that worked like the "sent" box in actual email systems.

Yes could be better to get there at once as if it works like that it probably makes whole mail mod just even more confusing.

Probably easiest would be to auto migrate db copying sent messages and 1-upping some storage version thing.

Of course there's fancy ways to do it like linked lists and such but I guess just retaining copy makes it easiest to maintain and understand which in turn reduces future mistakes. I'd say wasted storage space isn't really significant here.

Athozus commented:
Thanks for replies, I'll work on it, then I'll try to make a database with notes for messages deletion for each player, but the message is still there. Then it will be the occasion to add labels to sort messages.

@fluxionary , how your messages was not visible ? What was the values entered into the forms ?

fluxionary commented:
> @fluxionary , how your messages was not visible ? What was the values entered into the forms ?

apparently this isn't 100% consistent, but i haven't done much.

send a message:
![image](https://user-images.githubusercontent.com/25628292/221060162-1a9fd490-7d42-4945-aeeb-98d3c1e5319f.png)

i've got 3 messages in my sent box.
![image](https://user-images.githubusercontent.com/25628292/221060318-1512701e-7d7f-434f-a077-0d370a3a0d4d.png)

however, i can only "read" the first one - highlighting the other two and pressing "read" does nothing. none of the messages had any CCs or BCCs or anything weird.

Athozus commented:
Ok. Should I add a variable to keep the tab when the user go back ?

S-S-X commented:
> Should I add a variable to keep the tab when the user go back ?

That would be nice for UX, I'd say not really hard requirement for PR to get merged but still very nice if it keeps tab instead of always starting from first.

Important things IMO are that
* all sent messages can be read if they appear on list
* received messages not affected when sender removes / changes / updates / does whatever to message
* sent messages not affected when recipient removes / changes / updates / does whatever to message

That's core functionality for it to be useful and not too confusing, most other stuff is secondary and not necessarily required for MVP.

Athozus commented:
I was wondering how we could do this "global" mail storage. First, I think we need file names like `_name.json`, to not overwrite data from other versions (if a player was named `mails` or something like that).

Then, we can have those files ?
* `mails` : {"id", "sender", "to", "cc", "bcc", "time", "subject", "body"}
* `status` : {"id", "player", "action": "read/unread/delete"}

Athozus commented:
I was wondering how we could do this "global" mail storage. First, I think we need file names like `_name.json`, to not overwrite data from other versions (if a player was named `mails` or something like that).

Then, we can have those files ?
* `mails` : {"id", "sender", "to" : "player, @mail_list", "cc", "bcc", "time", "subject", "body"}
* `status` : {"id", "player", "status": "read/unread/deleted"}
* `contacts` : {"owner", "contact", "note"}
* `mail_lists` : {"id", "owner", "name", "desc"}
* `mail_lists_contacts` : {"id", "contact", "status": "to/cc/bcc"}

What do you think about it ? And, if it is adopted, how to transfer data from old templates to the new one ?

Also related to #3

S-S-X commented:
> First, I think we need file names like `_name.json`

```c
#define PLAYERNAME_ALLOWED_CHARS "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_"
```

I guess underscore wont work because it is allowed in player name but some `#`, `+`, `=`, `$` or `^` would probably work just fine. I think all mentioned are good for Windows too.

Or just upgrade db and have both sent and received in same per player JSON file, upgrading db to different isn't big issue anyway and can be done any time if it helps.
Mod storage could also work and you can have keys for per player sections like `mails.Sam.sent`, `mails.Sam.received`, there player name isn't issue because you can decide to add prefix to all storages instead of just for sent mails.

Most important thing with mail mod db migration is that process should never fail because of crap you throw at it, file based db for mails been in terrible state.
For example some time ago players were able to write their mail to any file they wanted, I think that was fixed though... maybe... but anyway for migration process there will be a lot of out of spec stuff and some panic fallback / automated repairing / validation is must.

fluxionary commented:
> I just finished a pre-implementation

some problems:
* mail which was received before this PR is inaccessible, you'll need to check for and migrate old mail to the new system
* i still can't read sent mail, or do anything with it. the buttons seem to act on the corresponding mail in my "received" mailbox.

Athozus commented:
>  mail which was received before this PR is inaccessible, you'll need to check for and migrate old mail to the new system

Yep, I don't know if I need to do a shell script, or integrate the old reading directly in the mail code. A script to transfer mails would be easy to do I think.

>  i still can't read sent mail, or do anything with it. the buttons seem to act on the corresponding mail in my "received" mailbox.

I don't understand why you get that, can you tell steps to reproduce pls ? It works perfectly for me.

S-S-X commented:
> Yep, I don't know if I need to do a shell script, or integrate the old reading directly in the mail code. A script to transfer mails would be easy to do I think.

No shell scripts. Use Lua mod code for migration, to keep it clean you can add new source file or even directory for it.
Upgrade should happen automatically when mod is loaded, old data can and probably should be left behind.

Just add documentation for cleanup, like generic step by step instructions.
For example: step 1. delete mail directory from world data directory 2. you really think there's more steps?

Athozus commented:
> No shell scripts. Use Lua mod code for migration, to keep it clean you can add new source file or even directory for it. Upgrade should happen automatically when mod is loaded,

Ok, I think this is easy to do that.

> old data can and probably should be left behind.

you mean that we don't care about old mails ?

> Just add documentation for cleanup, like generic step by step instructions. For example: step 1. delete mail directory from world data directory 2. you really think there's more steps?

So, a lua script or what ?

And, did you find issues ?

S-S-X commented:
> you mean that we don't care about old mails ?

No, I meant that you should not do any automated cleanup after successful migration. Just leave old stuff behind (acts as a disaster recovery backup guaranteed).
And add simple instructions how to destroy old data that isn't used anymore after migration: what server admin can do to cleanup old files that are not used anymore.

> So, a lua script or what ?

Yes, Lua script that will not be launched from command line but is integral part of the mail mod.
Conversion process, for server admin or minetest player, should be 2 steps:
* Update mod
* Launch minetest

And should **NOT** be anything that includes shell scripts, like not this:
* Update mod
* Run mail-db-migration.sh <-- this should not be part of the upgrade process
* Launch minetest

Of course if you'd like to provide shell script then just design mod part of the script, one that does actual conversion, in a way that you can easily wrap it in another script that allows usable command line interface. But this shouldn't be part of the mail mod.

Athozus commented:
Ok thanks for your explanations @S-S-X, I will do that quickly.

Concerning the new database itself, and other new features, did you find any issues like crashes/nonfunctional stuffs... ?

S-S-X commented:
> Concerning the new database itself, and other new features, did you find any issues like crashes/nonfunctional stuffs... ?

Have not read the code yet but will do that when I have time to concentrate properly for a few moments, possibly tomorrow but not 100% sure.
If it seems that basics work then probably wont require big changes anyway.

Version information for new database if you've not added it already:
One thing that would be good to add is database version number written in actual database, that makes it easier to check what has to be done for next update. True that it isn't strictly needed for first version but would be good to have.

Athozus commented:
> Version information for new database if you've not added it already:
One thing that would be good to add is database version number written in actual database, that makes it easier to check what has to be done for next update. True that it isn't strictly needed for first version but would be good to have.

Would you like to add a comment ? In Json files, in Lua ?

S-S-X commented:
> Would you like to add a comment ? In Json files, in Lua ?

Lua source files should contain mod version, for example `mail.version = 1234` in Lua and database version `"version": 1233` in JSON files.
Basic idea is that when mail mod is loaded it can read version written to storage and check if that version is older than installed mod version and start database upgrade if mod is updated but database isn't yet.

During startup, mod will check if database version is older than mod version. If it is then mod starts database upgrade and during that upgrade it also writes new version number to database.

This first upgrade does not have this information available so it wont exactly help with this upgrade but it will make it bit easier next time, also helps bit with debugging if something weird happens while updating mod.

fluxionary commented:
> > i still can't read sent mail, or do anything with it. the buttons seem to act on the corresponding mail in my "received" mailbox.
> 
> I don't understand why you get that, can you tell steps to reproduce pls ? It works perfectly for me.

start w/ a totally empty inbox and sent box. send a mail to another player. go to the sent box, and try to read the mail. nothing happens. send a mail from another player to the first account. on the first account, go to the sent box again, and click read. you will instead see the message you just received. click "delete", and it will delete the message you just received.

Athozus commented:
o_O @fluxionary you're right, I will fix that

Athozus commented:
@fluxionary this has been fixed, it was a problem with `selected_idxs`. I thought the `messages` key was ok for both tables, but in fact not, I had to create `inbox` and `sent`. I also worked to get the correct message id and extract the correct message from the correct id into `show_message()`. I had to create to `getPlayerMessages()`, which gives inbox+sent (so the old function is now moved to `getPlayerInboxMessages()`, and `getMessage()` without plural because we send the id and it returns the message table. I hope it works for you now.

fluxionary commented:
> @fluxionary this has been fixed

nice! things seem to work the way i'd expect now, as far as i've tested. 

i'd personally prefer that the "sent mail" feature be merged as-is, and mail-lists be put into a separate PR, though i'll leave that decision to you, as it might be a little tricky to split the code out at this point.

for the mail-lists, a few small issues:
* there's no way to add mail-lists to the recipient fields through the GUI, the way you can add recipients from the contact list
* i'm not sure what the point of having separate "to/cc/bcc" attributes for individuals in a mail list, the list is never "expanded" - when i receive a mail that was sent to a list, it just lists the recipient as the mail list.
* it'd be nice to be able to add people to the mail list from the contact list

S-S-X commented:
> i'd personally prefer that the "sent mail" feature be merged as-is, and mail-lists be put into a separate PR, though i'll leave that decision to you, as it might be a little tricky to split the code out at this point.

I would have exact same recommendation. It is a lot easier and faster to get multiple smaller pull requests reviewed that one large PR.

If significant stuff from one PR is required for another feature then I've often been simply basing one feature branch on top of another and in the end creating sequence of pull requests with information about dependencies. Easy to work on all things while keeping stuff in different branches, ofc requires a lot of rebasing but that's semi automatic anyway.

Athozus commented:
> nice! things seem to work the way i'd expect now, as far as i've tested.

Nice to hear that ! =)

> i'd personally prefer that the "sent mail" feature be merged as-is, and mail-lists be put into a separate PR, though i'll leave that decision to you, as it might be a little tricky to split the code out at this point.

Yep, me too, but as I had to made many stuff after noting that just sent messages cannot be functional alone => then new database, etc so I added the maillists btw.

> for the mail-lists, a few small issues:

Thanks for reporting those =)

>     * there's no way to add mail-lists to the recipient fields through the GUI, the way you can add recipients from the contact list

You mean, selecting contacts like clicking on To/Cc/Bcc buttons for adding receivers to messages ?

>     * i'm not sure what the point of having separate "to/cc/bcc" attributes for individuals in a mail list, the list is never "expanded" - when i receive a mail that was sent to a list, it just lists the recipient as the mail list.

Yep, in fact this is a feature I'd like to discuss, maybe you don't agree.
1. For each maillist, take the lower between position of the maillist and position of the player. Ex: to & cc -> cc ; bcc & to -> bcc ; to & to -> to
2. Then, take the higher of each recipient. Ex : a player is registered via one maillist in cc (via 1. script), and he's in manually written in to, so he's in to.

Global example : 
* @ml1 : player bcc
* message => to: @ml1, cc: player, bcc: --
Then the player receive the message from cc, because his maillist gives him bcc (bcc < to).

I hope it's clear for you

>     * it'd be nice to be able to add people to the mail list from the contact list

Ok, I'll try to do that.

Thanks for all =)

Athozus commented:
> I would have exact same recommendation. It is a lot easier and faster to get multiple smaller pull requests reviewed that one large PR.
> 
> If significant stuff from one PR is required for another feature then I've often been simply basing one feature branch on top of another and in the end creating sequence of pull requests with information about dependencies. Easy to work on all things while keeping stuff in different branches, ofc requires a lot of rebasing but that's semi automatic anyway.

@S-S-X : As I said to @fluxionary, I added all this stuff because first I had to add the new database to make the sent messages work. And, it was my mistake, I added the maillist moreover. Idk how to split the branches now, I hope you'll merge both =)

Now, I just need to make the lua script to upgrade database version.

S-S-X commented:
>  Idk how to split the branches now, I hope you'll merge both =)

Of course when things seems good sure this will get merged, it is just that reviews and testing with such big changes especially when it is multiple features might take a while.
If splitting is possible somehow then I'd strongly recommend doing it but not no need to worry, splitting isn't strictly required if it seems to be too much work.

Athozus commented:
> Of course when things seems good sure this will get merged, it is just that reviews and testing with such big changes especially when it is multiple features might take a while. If splitting is possible somehow then I'd strongly recommend doing it but not no need to worry, splitting isn't strictly required if it seems to be too much work.

Ok, thanks so =)

I will just work on version upgrade now, and then fix issues until you merge.

S-S-X commented:
> I will just work on version upgrade now, and then fix issues until you merge.

Sounds good 👍
Just comment here when database upgrade is ready and after that actual reviews can be done.

But also it would be best if you can avoid pushing any new features here on this branch and only complete new database model, storage upgrade and other stuff directly related to underlying mail database.

Best to make another branch if you want to add more features, update gui or anything not directly related to new database / storage. Also feel free to ask if you need any help with multiple feature branches / git workflows / rebasing your work.

Athozus commented:
> Sounds good +1 Just comment here when database upgrade is ready and after that actual reviews can be done.

Yep, I hope tomorrow would be ok.

> But also it would be best if you can avoid pushing any new features here on this branch and only complete new database model, storage upgrade and other stuff directly related to underlying mail database.

I won't add new features then, just finish this pull request.

> Best to make another branch if you want to add more features, update gui or anything not directly related to new database / storage. Also feel free to ask if you need any help with multiple feature branches / git workflows / rebasing your work.

Let's say for this pull request it's ok, for further I will split in many branches, as I would have liked to do initially.

Athozus commented:
Hi @S-S-X , @fluxionary ,

I finished migration scripts (thanks so SX recommendations). I let you try it, I hope it doesn't contain many issues, for my own I tried many times and it worked well.

About the mtt issue, no worry, it was just a `mkdir` missing, so it wasn't possible to write mail files.

I added the file `mail.info.json`, which contains currently only the value `dbversion`, which is used to compare with the version stored in mail code.

See you then =)

S-S-X commented on storage.lua:
```diff
@@ -1,17 +1,10 @@
-
-function mail.getMessages(playername)
-	local messages = mail.read_json_file(mail.getMailFile(playername))
+function mail.getMessages()
```
Consider this for better backwards compatibility (if returned data is still backwards compatible):
```suggestion
function mail.getMessages(playername)
	if (playername) then
		return mail.getPlayerMessages(playername)
	end
```

Athozus commented:
I pushed your commit @S-S-X , thanks, now looking for issues, if there are, and merge ?

S-S-X commented on api.lua:
```diff
@@ -116,3 +112,18 @@ function mail.send(...)
 		end
 	end
 end
+
+function mail.split(str, delimiter) -- flux split function
+    local rv = {}
+    local i, j = str:find(delimiter, nil, true)
+
+    while i do
+        table.insert(rv, str:sub(1, i - 1))
+        str = str:sub(j + 1)
+        i, j = str:find(delimiter, nil, true)
+    end
+
+    table.insert(rv, str)
+
+    return rv
+end
```
Seems like this should be removed and `string.split` used instead, for usage see next comment.
```suggestion
```

S-S-X commented on util/normalize.lua:
```diff
@@ -6,11 +6,15 @@ and add individual player names to recipient list
-        else
-            recipients[player_name] = true
+    for _, recipient_name in ipairs(order) do
+        if mail.split(recipient_name, "@")[1] == "" then -- in case of maillist
```
Use `string.split` instead of `mail.split`.
```suggestion
        if recipient_name:split("@")[1] == "" then -- in case of maillist
```

BuckarooBanzay commented:
i just tried this with the existing mail-set from pandorabox (in the `mails/${playername}.json` format) but somehow it didn't migrate my user :thinking: (or any other for that matter)

Haven't debugged it yet, did i miss something obvious to trigger this?

Athozus commented:
@BuckarooBanzay as agreed with @S-S-X , we keep old files (`BuckarooBanzay.json`) in case of error while migrating. So, now, you should see some files `mail.messages.json` in your `mails/` directory.

Else, could you send me a download link for your world, then I'll try it.

BuckarooBanzay commented:
> now, you should see some files mail.messages.json in your mails/ directory.

Yeah, my fault: the players i was testing weren't yet in the auth db, migration works now, i added a few commits though:
* re-enabled luacheck (you might want to look into the errors, there are plenty of unused vars and global var leaks)
* added a check for missing `to` fields in old storage formats
* re-enabled the `mtt` if branch (there are settingtypes now in the mod, you should be able to toggle the testing in the menu)

Correct me if i'm wrong:
* _all_ the mails are now in a single json file right?
* and: the "sent" mails are computed from the whole mail-set, correct?
* If so: wouldn't this be an issue with a larger player-base in the long run?

Athozus commented:
> Yeah, my fault: the players i was testing weren't yet in the auth db, migration works now, i added a few commits though:

Ok so, thanks

>     * re-enabled luacheck (you might want to look into the errors, there are plenty of unused vars and global var leaks)

I will look on it and fix errors, thx for adding this

>     * added a check for missing `to` fields in old storage formats

Thanks for that too, I didn't think about it

>     * re-enabled the `mtt` if branch (there are settingtypes now in the mod, you should be able to toggle the testing in the menu)

Thx for that, I had some problems with that but finally I installed it on I can run it directly on my pc.

>     * _all_ the mails are now in a single json file right?

Mails, yes, in `mail.messages.json`. Mails status (read/unread/deleted) are in `mail.status.json`. All different data are split into different files.

>     * and: the "sent" mails are computed from the whole mail-set, correct?

Yes, instead of looking at to/cc/bcc, it looks at the sender.

>     * If so: wouldn't this be an issue with a larger player-base in the long run?

I don't know, I will run a test (up to 100k would be good).

S-S-X commented:
> For the lot of messages check, I think I didn't wrote it correctly, so it lagged, but in fact I'm pretty sure it's not really a problem

This should not be automated test but one time performance check preferably against previous performance, if it is worse than before then probably should do something.

S-S-X commented on init.lua:
```diff
@@ -2,6 +2,9 @@ mail = {
 	-- api version
 	apiversion = 1.1,
```
API has changed so should probably update this version number.

To follow semantic versioning should go through API changes to find out what new version should be.

Best would be to somehow document / summarize API changes, documentation of course is good but for pull request I think writing out summary of changes as PR comments / possibly including in PR description is most helpful.

Athozus replied:
Ok, thx.

S-S-X replied:
This is still relevant, do you happen to have summary of API changes? (changed/removed/added function names, arguments and return values)
Of course I can also do this but it is a lot of API changes so I have to find some extra time for this.

I think for example mail.send arguments might not be compatible anymore but didn't check.

S-S-X commented on util/normalize.lua:
```diff
@@ -6,11 +6,15 @@ and add individual player names to recipient list
-        else
-            recipients[player_name] = true
+    for _, recipient_name in ipairs(order) do
+        if recipient_name:split("@")[1] == "" then -- in case of maillist
+            recipients[recipient_name] = true
```
Should this also check if mailing list exists (has at least 1 subscriber) and add to undeliverable if it does not? I mean similar how player check is done but for lists.

I've no idea just yet how mailing lists work but doing all recipient validation here feels like correct thing to do if possible.

Athozus replied:
Yes, I think this line was deleted, thx.

Athozus commented on mtt.lua:
```diff
@@ -1,14 +1,13 @@
-    mail.send("player1", "player2", "something", "blah")
+    mail.send({from = "player1", to = "player2", subject = "something", body = "blah"})
 
     -- check the receivers inbox
-    local list2 = mail.getMessages("player2")
+    local list2 = mail.getPlayerInboxMessages("player2")
```
Normally, it should works, I think @BuckarooBanzay made it.

S-S-X replied:
@BuckarooBanzay for existing tests I think there's 2 good options:
* Either restore old tests and add new style / new API as a new tests so that both are tested.
* OR add deprecation warning for old API functions. (I would prefer this if new stuff seems better)
  * and possibly just delete stuff that probably isn't used even if public...

Athozus commented:
Hi, based on your previous suggestions, and as the commit summary says :

* Now, when you send a message, the maillist is replaced by its players. Then, further players listed will not receive the message when they weren't in the maillist. (@fluxionary )
* Check if the maillist exists (@S-S-X )
* Now code written with the goal to be validated by luacheck (@BuckarooBanzay )
* Check missing fields on to/cc/bcc
* Remove unusable status in players lists, now the lists is basic "Player1, Player2, ..." (@fluxionary , @S-S-X ) -> extractMaillist() now reworked
* Avoid duplicates when editing more than 2 times a same maillist (ex : Player2 occures two times)
* Remove unused functions / use builtin functions

Hope it will be ok for you ; good evening =)

Athozus commented:
Good, thx =)

BuckarooBanzay commented:
@Athozus i got around to ~~butcher~~refactor the storage api surface:

* All the functions in `storage.lua` are now retrieving a mod-storage table and save to it after modification
* Tested it locally a bit but there might be further bugs :bug: 
* Moved and separated all the ui code to their own module (this has bugged me for far too long)
* Cleaned up most of the ui/state mess, there are still some issues but nothing big i think
* Removed all the migration code from v1 to v2 and left a note in the readme (can still be done with a commit before the changes)
* Mail-messages have a uuid as `id` field instead of the playername-timestamp combo (this ensures uniqueness)

Please let me know if my changes are alright for you

Athozus commented:
Hi, thanks for all the rework you did for me. I think that this storage would be better than the one I initially did, so thanks for that.

I'm gonna try it tomorrow but as checks are green it should be alright.

I just disagree one thing, concerning migrating from v1. For the ones still in v1 it is unfair, so can you amend the commit which delete it ? If there is a separated commit for that deletion, you can do (using `git log` to know the commits) :

```
git rebase -i previous-commit-hash-8-first-chars
```
`pick` --> `drop` at the line before your deletion commit.

Thanks for all anyway, I will check that more in details later.

S-S-X commented on storage.lua:
```diff
@@ -1,47 +1,194 @@
-function mail.getContactsFile(playername)
-	local saneplayername = string.gsub(playername, "[.|/]", "")
-	return mail.maildir .. "/contacts/" .. saneplayername .. ".json"
+function mail.get_storage_entry(playername)
+	local str = mail.storage:get_string(playername)
+	if str == "" then
+		-- new entry
+		return populate_entry()
+	else
+		-- deserialize existing entry
+		local e = minetest.parse_json(str)
+		return populate_entry(e)
+	end
 end
 
+function mail.set_storage_entry(playername, entry)
+	mail.storage:set_string(playername, minetest.write_json(entry))
+end
```
It would probably be good to add some static prefix to player storage keys so that there's clean keys for other stuff if needed later (without prefixes while possible it'll be annoying if one day you'd like to store some other stuff).

I think just adding any static prefix here should work without any other changes, like even just `"mail/" .. playername`.

BuckarooBanzay replied:
done :+1:

Athozus commented:
Hi, I fixed some little "mistakes" in the git tree. Sorry for the "co-authored" commits with me (on the last one), I don't know how it apparead, so a fix would be good but it's not a major problem I think. Else, I tried it and it works well, thx for the work. I just don't know where are messages stored, I used `ack` but nothing.

BuckarooBanzay commented:
tested this locally and with an existing server-setup (for the migrations) works well so far, merging this
i also added a quick-fix for the `unified_inventory` mod in the last commit ^

thanks for all the work :+1:

Athozus commented:
> tested this locally and with an existing server-setup (for the migrations) works well so far, merging this i also added a quick-fix for the `unified_inventory` mod in the last commit ^
> 
> thanks for all the work +1

Thanks ! Thank you for finally finishing the work of this very long merge request :D Thanks also to @S-S-X and @fluxionary who helped me during a long time ;)
